### DISCUSSION TITLE
Discussion: Variadic methods

### DISCUSSION THREAD

**Discussion by SamVanheer** [dotnet/csharplang#2330] [Language Ideas]
It is often necessary to write a method that can operate on a variable number of objects that also have a variable number of types.

For example, in a component system you may want to add multiple components of different types:
```
entity.AddComponents<Component1, Component2, Component3>();
```

It is not possible to write a method that can handle a variable number of types. The closest alternative is writing overloads for varying numbers of type parameters:
```
public void AddComponents<T0>()
where T0 : Component
{
    AddComponent<T0>();
}

public void AddComponents<T0, T1>()
where T0 : Component
where T1 : Component
{
    AddComponent<T0>();
    AddComponent<T1>();
}

//More overloads
```

I propose a solution that somewhat inverts generic methods:
```
public void AddComponents<variadic T>()
where T: Component
{
    foreach (var type in T)
    {
        AddComponent<type>();
    }
    //Compiler expands the loop to generate the same code as above
}
```

Instead of a generic method this method is more like a C++ template, instantiated by the compiler on demand and not a real generic method. Instead, a method is created that takes N non-generic parameters:
```
entity.AddComponents<Component1, Component2, Component3>();

//Not directly callable since multiple methods with this name could be generated
//Could be called when used as a delegate or acquired using Reflection
public void AddComponents();
```

In the case where only the generic types are variadic the compiler could optimize the generated methods by generating generic methods taking N generic parameters instead. Generic methods already support any operations that can be performed on the generic types, so there would be no need for a specific instance for each call to it.

At runtime these would be separate methods, just like the manually written overloads are.

The type of T would be a special compiler recognized type `VariadicTypes` that provides properties that are expanded by the compiler:
```
public readonly ref struct VariadicTypes
{
    public int Count { get; }

    public Type this[int index] { get; }

    public Enumerator GetEnumerator();

    public ref struct Enumerator
    {
        //Details omitted
    }
}
```

This type would be a stand-in used to act as an interface to the compiler generated code. `Count` is replaced with the number of parameters, indexed access is replaced with the actual type (which can be used with operators like `is` and `as`).

This would also work with parameters as well:
```
public void AddComponents(variadic Type types)
{
    foreach (var type in types)
    {
        AddComponent(type);
    }
    //Compiler expands the loop to generate this:
    AddComponent(type0);
    AddComponent(type1);
    AddComponent(type2);
}
```

Support for `ref` and `out` will be needed:
```
//Compiler recognizes relationship between generic parameter and regular parameter,
//and constrains the count to match
//(e.g. "incorrect number of parameters passed for variadic method AddComponents. 3 expected, 2 found")
public void AddComponents<variadic T>(variadic out T components)
{
    for (var i = 0; i < T.Length; ++i)
    {
        components[i] = AddComponent<T[i]>();
    }
}
```

The type of a variadic parameter is `VariadicParameters`, which looks a lot like `VariadicTypes`:
```
public readonly ref struct VariadicParameters
{
    public int Count { get; }

    public object this[int index] { get; }

    public Enumerator GetEnumerator();

    public ref struct Enumerator
    {
        //Details omitted
    }
}
```

As with `VariadicTypes`, this is an interface to the compiler generated code, so it would be possible to write the following:
```
public static string Format<variadic T>(string formatString, in variadic T parameters)
{
    //Actual formatting code would be more complex
    var builder = new StringBuilder();

    foreach (var parameter in parameters)
    {
        builder.Append(parameter.ToString());
    }

    return builder.ToString();
}
```

Which would expand to something like this:
```
public static string Format(string formatString, in int param0, in float param1, in string param2, in HeavyWeightStruct param3)
{
    var builder = new StringBuilder();

    builder.Append(param0.ToString());
    builder.Append(param1.ToString());
    builder.Append(param2.ToString());
    builder.Append(param3.ToString());

    return builder.ToString();
}
```

This completely eliminates the boxing overhead for value types.

It would also be possible to do this:
```
public List<string> ToStringList<variadic T>(variadic T parameters)
where T: struct
{
    var list = new List<T>(parameters.Count);

    foreach (var param in parameters)
    {
        //Compiler knows that param is IFormattable at compile time, no cast is performed
        if(param is IFormattable)
        {
            list.Add(param.ToString(null, CultureInfo.InvariantCulture));
        }
        else
        {
            list.Add(param.ToString());
        }
    }

    return list;
}
```

Which could be generated as:
```
public List<string> ToStringList(int param0, float param1, bool param2, char param3)
{
    var list = new List<T>(4);

    list.Add(param0.ToString(null, CultureInfo.InvariantCulture));
    list.Add(param1.ToString(null, CultureInfo.InvariantCulture));
    list.Add(param2.ToString());
    list.Add(param3.ToString());

    return list;
}
```

This should also be compatible with yield return:
```
public IEnumerable<TResult> FilterType<TResult, variadic T>(variadic T parameters)
{
    foreach (var param in parameters)
    {
        if (param is TResult)
        {
            yield return param;
        }
    }
}

//Prints:
//123
//45.67
//Does perform boxing to return as IFormattable
foreach (var formattable = FilterType<IFormattable>(123, 45.67f, true, 'a'))
{
    Console.WriteLine("{0}", formattable.ToString(null, CultureInfo.InvariantCulture));
}
```

And finally, it should be possible to write this, which could supersede params:
```
public static T[] ToArray<T>(in variadic T parameters)
{
    var result = new T[parameters.Count];

    for (var i = 0; i < parameters.Count; ++i)
    {
        result[i] = parameters[i];
    }

    return result;
}

//Used like:
var arrayOfThings = ToArray(ExpensiveThing1, ExpensiveThing2, ExpensiveThing3);

//Generated as:
public static ExpensiveThing[] ToArray(in ExpensiveThing param0, in ExpensiveThing param1, in ExpensiveThing param2)
{
    var result = new T[3];

    result[0] = param0;
    result[1] = param1;
    result[2] = param2;

    return result;
}
```

The compiler could then inline the method entirely, allowing it to be omitted from the compiled assembly.

In effect this would allow you to trade in memory usage in terms of methods loaded at runtime with better performance, avoiding boxing and not having to manually write overloads, which eliminates copy paste errors.

**Comment by HaloFour** [comment]
This seems like this would involve some fairly complicated code generation and possibly an entire extension of the C# language to support properly, and it would still be very limited in that the template couldn't cross assembly boundaries.

**Reply by Xyncgas** [reply]
> and it would still be very limited in that the template couldn't cross assembly boundaries
You're right.

We have source generators at this moment, if source generators can cross assembly boundaries (it's just normal c# codes), c++ templates can be possible

**Reply by HaloFour** [reply]
Source generators can emit new code but they cannot do anything with existing code or assemblies.  They certainly cannot emulate C++ templates based on generic definitions in other assemblies.  The source generator neither has access to the original source of that assembly, nor can it emit anything to replace anything within that assembly.

**Comment by CyrusNajmabadi** [comment]
> Which would expand to something like this:

Can you explain how that expansion would work?  How would the compiler even know the original body of hte method contained:

```
{
    //Actual formatting code would be more complex
    var builder = new StringBuilder();

    foreach (var parameter in parameters)
    {
        builder.Append(parameter.ToString());
    }

    return builder.ToString();
}
```

?  Method bodies are opaque to the language.  They may not even exist at compile time due to things like ref-assemblies.  So it's unclear ot me how this could work.

**Comment by SamVanheer** [comment]
@HaloFour Perhaps variadic methods could be stored in the compiled assembly so that it can be used in other assemblies?

@CyrusNajmabadi I assume the compiler has access to the code it's compiling? I'm not that familiar with the specifics of the C# compiler so i don't know if there are limitations that would prevent this sort of thing from working.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I assume the compiler has access to the code it's compiling? I'm not that familiar with the specifics of the C# compiler so i don't know if there are limitations that would prevent this sort of thing from working.

It does not.  A core part of the design of .net is that you can just compile again pure signatures of things.  So, for example, there are 'ref assemblies' that literally contain no code whatsoever. All they have are the type-definitions and member signatures that will be in the final assembly.

**Comment by Joe4evr** [comment]
> It is not possible to write a method that can handle a variable number of types.

Isn't the upcoming `Variant` type being made for exactly this purpose? [Proposal doc](https://github.com/dotnet/csharplang/blob/master/proposals/format.md#variant), [CoreFX Lab PR](https://github.com/dotnet/corefxlab/pull/2595).

**Comment by SamVanheer** [comment]
> 
> 
> > It is not possible to write a method that can handle a variable number of types.
> 
> Isn't the upcoming `Variant` type being made for exactly this purpose? [Proposal doc](https://github.com/dotnet/csharplang/blob/master/proposals/format.md#variant), [CoreFX Lab PR](https://github.com/dotnet/corefxlab/pull/2595).

Looks like it, but `Variant` in that proposal only supports a known subset of value types. It employs type erasure, and requires switch(type) constructs to handle anything type-dependent.

Even with `Variant` it is not possible to write a method that can take a variable number of generic types.

**Comment by yaakov-h** [comment]
I believe `Variant` is for a fixed number of varying types, not a variable number of types/arguments.

**Comment by TonyValenti** [comment]
I think you are able to get pretty close to that with:
https://github.com/dotnet/csharplang/issues/2230

**Comment by dsaf** [comment]
@SamVanheer 

I am not sure the sample justifies this feature. You need to give concrete examples of ```Component``` and ```AddComponent```.

> It is often necessary to write a method that can operate on a variable number of objects that also have a variable number of types.

In which case you either treat them as ```object```s or introduce a new generic contract that allows to treat them in the same way.

**Comment by scalablecory** [comment]
Not sure how I feel about this specific implementation, but I'd love this feature for anything involving generic tuple code.

**Comment by deeprobin** [comment]
Instead of using `variadic` we could use `params`...

something like that:
```cs
public static string Format<params T>(params T parameters) {
}
```
